source/opt/decoration_manager.cpp

```diff
@@ -260,16 +311,25 @@ void DecorationManager::ForEachDecoration(
 void DecorationManager::CloneDecorations(
     uint32_t from, uint32_t to, std::function<void(ir::Instruction&, bool)> f) {
   assert(f && "Missing function parameter f");
-  auto const decoration_list = id_to_decoration_insts_.find(from);
+  const auto decoration_list = id_to_decoration_insts_.find(from);
   if (decoration_list == id_to_decoration_insts_.end()) return;
-  for (ir::Instruction* inst : decoration_list->second) {
+  for (ir::Instruction* inst : decoration_list->second.direct_decorations) {
+    // simply clone decoration and change |target-id| to |to|
+    std::unique_ptr<ir::Instruction> new_inst(inst->Clone(module_->context()));
+    new_inst->SetInOperand(0, {to});
+    id_to_decoration_insts_[to].direct_decorations.push_back(new_inst.get());
```

Comment:
Doing the same hash lookup for every iteration.  Might want to move that out of the loop.

Reply:
I’ll store a reference to |to|’s `TargetData` outside of the loop, and use it in this loop and in others further down in the function.